src/libraries/Common/src/System/Reflection/Metadata/TypeNameParser.cs

```diff
@@ -0,0 +1,314 @@
+            {
+#if SYSTEM_PRIVATE_CORELIB
+                // backward compat: throw FileLoadException for non-empty invalid strings
+                if (!_throwOnError && _inputString.TrimStart().StartsWith(","))
+                {
+                    return null;
+                }
+                throw new IO.FileLoadException(SR.InvalidAssemblyName, _inputString.ToString());
+#else
+                return null;
+#endif
+            }
+
+            // No matter what was parsed, the full name string is allocated only once.
+            // In case of generic, nested, array, pointer and byref types the full name is allocated
+            // when needed for the first time .
+            string fullName = fullTypeName.ToString();
+
+            TypeName? declaringType = GetDeclaringType(fullName, nestedNameLengths, assemblyName);
+            TypeName result = new(fullName, assemblyName, declaringType: declaringType);
+            if (genericArgs is not null)
+            {
+                result = new(fullName: null, assemblyName, elementOrGenericType: result, declaringType, genericArgs);
+            }
+
+            if (previousDecorator != default) // some decorators were recognized
+            {
+                while (TryParseNextDecorator(ref capturedBeforeProcessing, out int parsedModifier))
+                {
+                    result = new(fullName: null, assemblyName, elementOrGenericType: result, rankOrModifier: (sbyte)parsedModifier);
+                }
+            }
+
+            return result;
+        }
+
+        /// <returns>false means the input was invalid and parsing has failed. Empty input is valid and returns true.</returns>
+        private bool TryParseAssemblyName(ref AssemblyName? assemblyName)
+        {
+            ReadOnlySpan<char> capturedBeforeProcessing = _inputString;
+            if (TryStripFirstCharAndTrailingSpaces(ref _inputString, ','))
+            {
+                if (_inputString.IsEmpty)
+                {
+                    _inputString = capturedBeforeProcessing; // restore the state
+                    return false;
+                }
+
+                // The only delimiter which can terminate an assembly name is ']'.
+                // Otherwise EOL serves as the terminator.
+                int assemblyNameLength = (int)Math.Min((uint)_inputString.IndexOf(']'), (uint)_inputString.Length);
+                ReadOnlySpan<char> candidate = _inputString.Slice(0, assemblyNameLength);
+                AssemblyNameParser.AssemblyNameParts parts = default;
+
+                if (!AssemblyNameParser.TryParse(candidate, ref parts))
+                {
+                    return false;
+                }
+
+                assemblyName = new AssemblyName();
+#if SYSTEM_PRIVATE_CORELIB
+                assemblyName.Init(parts);
+#else
+                assemblyName.Name = parts._name;
+                assemblyName.CultureName = parts._cultureName;
```

Comment:
The `AssemblyName.CultureName` property setter is not safe when given potentially hostile input.

Reply:
Excellent catch! I was unaware of the fact that it allocates the culture.

https://github.com/dotnet/runtime/blob/07648642d345087d344bc8b7b9b3ea7c8793bbea/src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs#L83

Reply:
@GrabYourPitchforks I've created a new proposal based on this feedback: https://github.com/dotnet/runtime/issues/100867